python_action/clang_format_xml.py

```diff
@@ -0,0 +1,126 @@
+from . import GlobalParser, get_line_cnt_from_cols
+
+class FormatReplacement:
+    """An object representing a single replacement.
+
+    Attributes:
+        cols (int): The columns number of where the suggestion starts on the line
+        null_len (int): The number of bytes removed by suggestion
+        text (str): The `bytearray` of the suggestion
+    """
+
+    def __init__(self, cols: int, null_len: int, text: str) -> None:
+        """
+        Args:
+            cols: The columns number of where the suggestion starts on the line
+            null_len: The number of bytes removed by suggestion
+            text: The `bytearray` of the suggestion
+        """
+        self.cols = cols
+        self.null_len = null_len
+        self.text = text
+
+    def __repr__(self) -> str:
+        return (
+            f"<FormatReplacement at cols {self.cols} removes {self.null_len} bytes"
+            f" adds {len(self.text)} bytes>"
+        )
+
+class FormatReplacementLine:
+    """An object that represents a replacement(s) for a single line.
+
+    Attributes:
+        line (int): The line number of where the suggestion starts
+        replacements (list): A list of
+            [`FormatReplacement`][python_action.clang_format_xml.FormatReplacement]
+            object(s) representing suggestions.
+    """
+
+    def __init__(self, line_numb: int):
+        """
+        Args:
+            line_numb: The line number of about the replacements
+        """
+        self.line = line_numb
+        self.replacements = []
+
+    def __repr__(self):
+        return (
+            f"<FormatReplacementLine @ line {self.line} "
+            f"with {len(self.replacements)} replacements>"
+        )
+
+
+class XMLFixit:
+    """A single object to represent each suggestion.
+
+    Attributes:
+        filename (str): The source file that the suggestion concerns.
+        replaced_lines (list): A list of
+            [`FormatReplacementLine`][python_action.clang_format_xml.FormatReplacementLine]
+            representing replacement(s) on a single line.
+    """
+
+    def __init__(self, filename: str):
+        """
+        Args:
+            filename: The source file's name for which the contents of the xml
+                file exported by clang-tidy.
+        """
+        self.filename = filename.replace(os.sep, "/")
+        self.replaced_lines = []
+
+    def __repr__(self) -> str:
+        return (
+            f"<XMLFixit with {len(self.replaced_lines)} lines of "
+            f"replacements for {self.filename}>"
+        )
+
+
+def parse_format_replacements_xml(src_filename: str):
+    """Parse XML output of replacements from clang-format. Output is saved to
+    [`format_advice`][python_action.__init__.GlobalParser.format_advice].
+
+    Args:
+        src_filename: The source file's name for which the contents of the xml
+            file exported by clang-tidy.
+    """
+    tree = ET.parse("clang_format_output.xml")
```

Comment:
This is used to parse XML output from clang-format. But it is a experimental feature that is disabled because parsing XML is only needed when aggregating suggestions from clang-tidy for diff comments.

I'm firmly against this suggestion
1. we are not actually parsing XML
2. the "Code Inspector - Code Review" suggestion (see below) recommends installing a 3<sup>rd</sup>-party lib that would slow down the action's setup process